File: java/src/MaxMatching.java
import java.util.*;

public class MaxMatching {

	static boolean findPath(boolean[][] graph, int u1, int[] matching, boolean[] vis) {
		vis[u1] = true;
		for (int v = 0; v < matching.length; ++v) {
			int u2 = matching[v];
			if (graph[u1][v] && (u2 == -1 || !vis[u2] && findPath(graph, u2, matching, vis))) {
				matching[v] = u1;
				return true;
			}
		}
		return false;
	}

	public static int maxMatching(boolean[][] graph) {
		int n1 = graph.length;
		int n2 = n1 == 0 ? 0 : graph[0].length;
		int[] matching = new int[n2];
		Arrays.fill(matching, -1);
		int matches = 0;
		for (int u = 0; u < n1; u++)
			if (findPath(graph, u, matching, new boolean[n1]))
				++matches;
		return matches;
	}

	// Usage example
	public static void main(String[] args) {
		int res = maxMatching(new boolean[][] { { true, true }, { true, false } });
		System.out.println(2 == res);
	}
}

File: python/max_matching.py
def max_matching(graph, n2):
    n1 = len(graph)
    matching = [-1] * n2
    matches = 0
    for u in range(n1):
        if find_path(graph, u, matching, [False] * n1):
            matches += 1
    return matches


def find_path(graph, u1, matching, vis):
    vis[u1] = True
    for v in range(len(matching)):
        u2 = matching[v]
        if graph[u1][v] and (u2 == -1 or not vis[u2] and find_path(graph, u2, matching, vis)):
            matching[v] = u1
            return True
    return False


def test():
    graph = [[False] * 2 for _ in range(2)]
    graph[0][1] = True
    graph[1][0] = True
    graph[1][1] = True
    assert 2 == max_matching(graph, 2)


test()
